### PR TITLE
fix: add custom metric from custom dimension

### PR DESCRIPTION
**Pull request by rephus**
Closes: https://github.com/lightdash/lightdash/issues/9931
### Description:
image

[Screencast from 15-05-24 09:40:51.webm](https://github.com/lightdash/lightdash/assets/1983672/b18f5412-1ae2-4e75-8d7c-14e85da6ef7c)

### Reviewer actions

- [ ] I have manually tested the changes in the preview environment
- [ ] I have reviewed the code
- [ ] I understand that "request changes" will block this PR from merging

### CODE REVIEW

File: packages/frontend/src/components/Explorer/CustomMetricModal/utils/index.ts [hunk 1]
```diff
@@ -73,9 +76,10 @@ const getCustomMetricDescription = (
     }`;
 
 const getTypeOverridesForAdditionalMetric = (
-    item: Dimension | AdditionalMetric,
+    item: Dimension | AdditionalMetric | CustomDimension,
     type: MetricType,
 ): Partial<AdditionalMetric> | undefined => {
+    //TODO handle custom dimensions based on dimensionType
     if (!isDimension(item)) return;
```
Comment (rephus) [inline packages/frontend/src/components/Explorer/CustomMetricModal/utils/index.ts:83]: I'm not sure if we need to change this

File: packages/frontend/src/components/Explorer/ExploreTree/TableTree/Tree/TreeSingleNodeActions.tsx [hunk 2]
```diff
@@ -176,7 +178,41 @@ const TreeSingleNodeActions: FC<Props> = ({
                     </Menu.Item>
                 )}
 
-                {customMetrics.length > 0 && isDimension(item) ? (
+                {isCustomDimension(item) && (
```
Comment (rephus) [inline packages/frontend/src/components/Explorer/ExploreTree/TableTree/Tree/TreeSingleNodeActions.tsx:181]: moved above, so custom metrics appear below

File: packages/frontend/src/components/Explorer/ExploreTree/TableTree/Tree/TreeSingleNodeActions.tsx [hunk 3]
```diff
@@ -176,7 +178,41 @@ const TreeSingleNodeActions: FC<Props> = ({
...
+                            }}
+                        >
+                            Remove custom dimension
+                        </Menu.Item>
+                    </>
+                )}
+
+                {customMetrics.length > 0 &&
+                (isDimension(item) ||
+                    (isCustomDimension(item) &&
```
Comment (rephus) [inline packages/frontend/src/components/Explorer/ExploreTree/TableTree/Tree/TreeSingleNodeActions.tsx:214]: show custom metric submenu for customDimensions

File: packages/frontend/src/components/Explorer/CustomMetricModal/index.tsx [hunk 4]
```diff
@@ -145,13 +146,14 @@ export const CustomMetricModal = () => {
     useEffect(() => {
         if (!item || !customMetricType) return;
 
-        if (item.label && customMetricType) {
+        const label = isCustomDimension(item) ? item.name : item.label;
```
Comment (ZeRego) [inline packages/frontend/src/components/Explorer/CustomMetricModal/index.tsx:149]: (non blocker) we can use the util function `getItemLabelWithoutTableName()`

File: packages/frontend/src/components/Explorer/ExploreTree/TableTree/Tree/TreeSingleNodeActions.tsx [hunk 5]
```diff
@@ -94,10 +94,12 @@ const TreeSingleNodeActions: FC<Props> = ({
         (context) => context.actions.toggleCustomDimensionModal,
     );
 
-    const customMetrics = useMemo(
-        () => (isDimension(item) ? getCustomMetricType(item.type) : []),
-        [item],
-    );
+    const customMetrics = useMemo(() => {
+        if (isCustomDimension(item) && item.type === CustomDimensionType.SQL) {
```
Comment (ZeRego) [inline packages/frontend/src/components/Explorer/ExploreTree/TableTree/Tree/TreeSingleNodeActions.tsx:98]:
```suggestion
        if (isCustomSqlDimension(item)) {
```

File: packages/frontend/src/components/Explorer/ExploreTree/TableTree/Tree/TreeSingleNodeActions.tsx (hunk 3 above)
Comment (ZeRego) [reply packages/frontend/src/components/Explorer/ExploreTree/TableTree/Tree/TreeSingleNodeActions.tsx:214]: Not sure why we need all these checks when `customMetrics` is going to be empty in those cases. Should we just have `customMetrics.length > 0` ?

File: packages/frontend/src/components/Explorer/CustomMetricModal/utils/index.ts (hunk 1 above)
Comment (ZeRego) [reply packages/frontend/src/components/Explorer/CustomMetricModal/utils/index.ts:82]: fixed

File: packages/frontend/src/components/Explorer/ExploreTree/TableTree/Tree/TreeSingleNodeActions.tsx (hunk 3 above)
Comment (ZeRego) [reply packages/frontend/src/components/Explorer/ExploreTree/TableTree/Tree/TreeSingleNodeActions.tsx:214]: ok, we need the `item` typing to access properties. Changed this to use `isCustomSqlDimension`.

**Review by ZeRego — APPROVED** [review]

**Comment by rephus** [comment]
Thanks for the changes @ZeRego

**Pull request merged**